core/db/db.go

```diff
@@ -52,59 +60,25 @@ type OpenchainDB struct {
 	StateDeltaCF *gorocksdb.ColumnFamilyHandle
 	IndexesCF    *gorocksdb.ColumnFamilyHandle
 	PersistCF    *gorocksdb.ColumnFamilyHandle
+	dbState      dbState
+	mux          sync.Mutex
 }
 
 var openchainDB *OpenchainDB
-var isOpen bool
+var once sync.Once
 
-// CreateDB creates a rocks db database
-func CreateDB() error {
-	dbPath := getDBPath()
-	dbLogger.Debug("Creating DB at [%s]", dbPath)
-	missing, err := dirMissingOrEmpty(dbPath)
-	if err != nil {
-		return err
-	}
-
-	if !missing {
-		return fmt.Errorf("db dir [%s] already exists", dbPath)
-	}
-	err = os.MkdirAll(path.Dir(dbPath), 0755)
-	if err != nil {
-		dbLogger.Error("Error calling  os.MkdirAll for directory path [%s]: %s", dbPath, err)
-		return fmt.Errorf("Error making directory path [%s]: %s", dbPath, err)
-	}
-	opts := gorocksdb.NewDefaultOptions()
-	defer opts.Destroy()
-	opts.SetCreateIfMissing(true)
-
-	db, err := gorocksdb.OpenDb(opts, dbPath)
-	if err != nil {
-		return err
-	}
-
-	defer db.Close()
-
-	dbLogger.Debug("DB created at [%s]", dbPath)
-	return nil
+// Create create an openchainDB instance
+func Create() *OpenchainDB {
+	return &OpenchainDB{dbState: closed}
 }
 
-// GetDBHandle returns a handle to OpenchainDB
+// GetDBHandle get an opened openchainDB singleton
 func GetDBHandle() *OpenchainDB {
-	var err error
-	if isOpen {
-		return openchainDB
-	}
-
-	err = createDBIfDBPathEmpty()
-	if err != nil {
-		panic(fmt.Sprintf("Error while trying to create DB: %s", err))
-	}
+	once.Do(func() {
+		openchainDB = Create()
+	})
+	openchainDB.Open()
```

Comment:
@manish-sethi I am not quite sure we can cache the function call because I found some codes such as test will close db and open again. So the open method should be invoked every time.Is that right?